File: cppsample/src/cleantests/cpp/task_storage.cpp
#include "task_storage.h"

void TaskStorage::add(Task task) {
    systemTasks_.push_back(task);
}

void TaskStorage::remove(Task task) {
    std::vector<Task>::iterator it = systemTasks_.begin();
    while (it != systemTasks_.end())
    {
        if (it->id_ == task.id_ && it->group_ == task.group_)
        {
            systemTasks_.erase(it++);
        }
        else
        {
            ++it;
        }
    }
}

void TaskStorage::clear() {
    systemTasks_.clear();
}

bool TaskStorage::tasksAvailable() {
    return !systemTasks_.empty();
}

std::vector<Task> TaskStorage::getTasks() {
    return systemTasks_;
}

std::vector<Task> TaskStorage::getTasksInGroup(std::string group) {
    std::vector<Task> tasksInGroup;
    std::vector<Task>::iterator it = systemTasks_.begin();
    while (it != systemTasks_.end())
    {
        if (it->group_ == group)
        {
            tasksInGroup.push_back(*it);
        }
    }
    return tasksInGroup;
}

File: javasample/src/main/java/cleanTests/TaskStorage.java
package cleanTests;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class TaskStorage {

    private List<Task> systemTasks = new ArrayList<>();

    public boolean tasksAvailable() {
        return !systemTasks.isEmpty();
    }

    public TaskStorage add(Task task) {
        systemTasks.add(task);
        return this;
    }

    public TaskStorage remove(Task task) {
        systemTasks.remove(task);
        return this;
    }

    public TaskStorage clear() {
        systemTasks.clear();
        return this;
    }

    public List<Task> getTasks() {
        return new ArrayList<>(systemTasks);
    }

    public List<Task> getTasksInGroup(String group) {
        List<Task> tasks =  systemTasks.stream()
                .filter( t -> t.group.equalsIgnoreCase(group))
                .collect(Collectors.toList());
        return tasks;
    }

}
